Study  -> Electronic Orders - React Implementation

sagarkhandagre998 opened the pull request:
🎯 Overview
Implemented complete React-based Study Electronic Orders functionality, eliminating redirects to old JSP UI. Users can now search, view, reject, and process electronic orders entirely within the React interface.

## 🔧 Technical Implementation
Implementation Steps

1. **Database Changes**
   - Added `hide_in_new_ui` column to `menu` table
   - Created React menu entry with visibility flags

2. **Backend REST API**
   - Created `RestStudyElectronicOrdersController`
   - Added `GET /rest/StudyElectronicOrders` for search
   - Added `POST /rest/rejectStudyElectronicOrder` for rejection
   - Added `setSearchType(String)` overload for parameter binding

3. **Menu Filtering**
   - Updated `MenuUtil` to filter by `hide_in_new_ui` flag
   - Modified `MenuController` to return filtered menus

4. **React Components**
   - Created `StudyElectronicOrders.js`, `StudyEOrderSearch.js`, `StudyEOrder.js`
   - Added route `/StudyElectronicOrders` in `App.js`

5. **Features**
   - Search by patient ID, date range, or facility
   - Reject modal with reason/notes (updates status to "Cancelled")
   - Edit redirects to React order entry
   - Disabled buttons for rejected orders

## 🧪 Testing
- **Cypress E2E Tests**: 6/6 passing
  - Page load, search, reject modal, reject workflow, status display, edit redirect
- **Manual Testing**: Verified with test data and database validation

## 📋 User Flow
1. Navigate to Order → Study → Electronic Orders
2. Search by patient code or date range
3. View results in table
4. Reject orders via modal or Edit/Process via React form
5. Rejected orders show disabled state

[Screencast from 2025-12-29 15-22-35.webm](https://github.com/user-attachments/assets/9f669e4a-7116-48c1-a2c0-d1ae1ebc6744)

sagarkhandagre998 committed:
Implemented Study -> ElectronicOrders react component

sagarkhandagre998 committed:
apply formating

vsvishalsharma commented on frontend/src/components/studyElectronicOrder/StudyEOrderSearch.js:
```diff
@@ -0,0 +1,355 @@
+  }, []);
+
+  const handleElectronicOrders = (response) => {
+    console.log("Initial response:", response);
```
@sagarkhandagre998 remove console logs from the code

sagarkhandagre998 replied:
Sorry I forgot to remove that, Sure I'll remove that.

vsvishalsharma commented on frontend/src/components/studyElectronicOrder/StudyEOrderSearch.js:
```diff
@@ -0,0 +1,355 @@
+  }
+
+  const parseEOrders = (response) => {
+    console.log("parseEOrders response:", response);
```
here as well

vsvishalsharma commented on frontend/src/components/studyElectronicOrder/StudyEOrderSearch.js:
```diff
@@ -0,0 +1,355 @@
+
+    if (eOrderRef?.current) {
+      window.scrollTo({
+        top: eOrderRef.current.offsetTop - 50,
```
@sagarkhandagre998 what is 50 here shall be a named constant

vsvishalsharma commented on frontend/src/components/studyElectronicOrder/StudyEOrderSearch.js:
```diff
@@ -0,0 +1,355 @@
+  };
+
+  function searchByIdentifier() {
+    const params = new URLSearchParams({
```
@sagarkhandagre998 This pattern is repeated three times. Should be extracted to a single function.

sagarkhandagre998 replied:
Sure

vsvishalsharma commented on frontend/src/components/studyElectronicOrder/StudyEOrderSearch.js:
```diff
@@ -0,0 +1,355 @@
+                  hasIconOnly
+                  id="loadpreviousresults"
+                  onClick={loadPreviousResultsPage}
+                  disabled={previousPage != null ? false : true}
```
this should be disabled={previousPage === null} or just disabled={!previousPage}

vsvishalsharma commented on frontend/src/components/studyElectronicOrder/StudyEOrder.js:
```diff
@@ -0,0 +1,420 @@
+                        (cell) => cell.info.header === "status",
+                      );
+                      const isCancelled =
+                        statusCell?.value === "Cancelled" ||
```
Checking for both spellings suggests data inconsistency. This should be normalized on the backend or use case-insensitive comparison

vsvishalsharma commented on frontend/src/components/studyElectronicOrder/StudyEOrder.js:
```diff
@@ -0,0 +1,420 @@
+      formData,
+      (status) => {
+        setIsSubmitting(false);
+        if (status === 200) {
```
@sagarkhandagre998 what about sratus 204 and 201 they are also a valid status code how are you handling that

vsvishalsharma commented on frontend/src/components/studyElectronicOrder/StudyEOrder.js:
```diff
@@ -0,0 +1,420 @@
+                                        handleEdit(
+                                          row.cells.find(
+                                            (c) =>
+                                              c.info.header ===
```
can you also create a constant for these info.header so that it is maintanable in long term

sagarkhandagre998 commented:
@vsvishalsharma Thanks for the review, I'll try to resolve Asap.

vsvishalsharma commented:
@sagarkhandagre998 once done with it please mark the comments resolved for better tracking

sagarkhandagre998 committed:
Fixing studyEorderSearch.js as recommanded

sagarkhandagre998 committed:
Fixing StudyEorder.js as recommanded

vsvishalsharma commented on src/main/java/org/openelisglobal/menu/util/MenuUtil.java:
```diff
@@ -80,6 +80,35 @@ public static List<MenuItem> getMenuTree() {
+        List<MenuItem> filtered = new ArrayList<>();
+
+        for (MenuItem menuItem : menuItems) {
+            Menu menu = menuItem.getMenu();
```
Potential NullPointerException
```
    Menu menu = menuItem.getMenu();
    if (menu.getIsActive() && !menu.isHideInNewUI()) {
```
No null checks:


- What if menuItems is null?
- What if menuItem.getMenu() returns null?
- What if menuItem.getChildMenus() returns null?

vsvishalsharma commented on src/main/java/org/openelisglobal/menu/util/MenuUtil.java:
```diff
@@ -80,6 +80,35 @@ public static List<MenuItem> getMenuTree() {
+            // Skip menus that should be hidden in new UI
+            if (menu.getIsActive() && !menu.isHideInNewUI()) {
+                MenuItem filteredItem = new MenuItem();
+                filteredItem.setMenu(menu);
```
Shares the same Menu object reference. If the filtered tree is modified, it affects the original tree.

vsvishalsharma commented on src/main/java/org/openelisglobal/menu/util/MenuUtil.java:
```diff
@@ -80,6 +80,35 @@ public static List<MenuItem> getMenuTree() {
+                MenuItem filteredItem = new MenuItem();
+                filteredItem.setMenu(menu);
+                // Recursively filter children
+                filteredItem.setChildMenus(filterMenuForReactUI(menuItem.getChildMenus()));
```
If getChildMenus() returns null, this will throw NPE in the recursive call

vsvishalsharma commented:
@sagarkhandagre998 few comments regarding null pointer exception else this LGTM

sagarkhandagre998 commented:
> @sagarkhandagre998 few comments regarding null pointer exception else this LGTM

Sure I'll Fix that.

Agaba-derrick commented:
@sagarkhandagre998 are you able to name the issue you are addressing , I  would recommend we always keep the OE pull request template for consistency !

sagarkhandagre998 commented:
> @sagarkhandagre998 are you able to name the issue you are addressing , I would recommend we always keep the OE pull request template for consistency !

This is a new feature, not listed in the issue dashboard. Should I create ?

sagarkhandagre998 committed:
fixing npe and ref obj as recommanded

vsvishalsharma commented on frontend/cypress/e2e/studyElectronicOrders.cy.js:
```diff
@@ -0,0 +1,273 @@
+      'docker exec openelisglobal-database psql -U clinlims -d clinlims -c "' +
+        "INSERT INTO clinlims.person (id, first_name, last_name, lastupdated) " +
+        "VALUES (999, 'Test', 'Patient', NOW()) ON CONFLICT (id) DO NOTHING; " +
+        "INSERT INTO clinlims.patient (id, person_id, national_id, external_id, birth_date, gender, lastupdated) " +
```
@sagarkhandagre998 is it really neccessary to add the data in db can we not use mock data for ui tests

sagarkhandagre998 replied:
You are correct, Thanks for the feedback! I used database setup for this initial implementation to validate the entire end-to-end flow while learning the feature. I agree that `cy.intercept()` with mock data would be better for pure UI testing - I will refactor it.

vsvishalsharma commented on frontend/cypress/e2e/studyElectronicOrders.cy.js:
```diff
@@ -0,0 +1,273 @@
+    // Clean up test data
+    cy.exec(
+      'docker exec openelisglobal-database psql -U clinlims -d clinlims -c "' +
+        "DELETE FROM clinlims.electronic_order WHERE patient_id = 999; " +
```
here as well we will not be required to run these commands if we use mock data for the ui functionss

vsvishalsharma commented on frontend/cypress/e2e/studyElectronicOrders.cy.js:
```diff
@@ -0,0 +1,273 @@
+    cy.get("#reject-note").should("exist");
+
+    // Verify primary Reject button is disabled without selecting reason
+    cy.get(".cds--modal-footer")
```
we can define a slector and use it here if possible to make the test more reliable

vsvishalsharma commented on frontend/cypress/e2e/studyElectronicOrders.cy.js:
```diff
@@ -0,0 +1,273 @@
+  });
+
+  it("Edit button redirects to order entry page", () => {
+    cy.intercept("GET", "/api/OpenELIS-Global/rest/StudyElectronicOrders*").as(
```
also this intercept method is not required in my opinion but if you want to keep it its repeatative make a const better to remove it

sagarkhandagre998 commented:
@vsvishalsharma  I think it's perfectly fine now, I'll do the same approach for study -> initial and double entry also.

sagarkhandagre998 commented:
@mozzy11  Please take a look.

sagarkhandagre998 committed:
fixing ui-test to use mock data

sagarkhandagre998 committed:
Created 3 constants at the top and Replaced all repetitive mock data** with these constants

sagarkhandagre998 committed:
Fix merge conflict in base.xml

Agaba-derrick commented:
@sagarkhandagre998  thanks .,are you able to rebase this !

sagarkhandagre998 commented:
> @sagarkhandagre998 thanks .,are you able to rebase this !

Welcome.

sagarkhandagre998 commented:
@pmanko Kindly look into this pr also.

pmanko reviewed (dismissed):
$(cat <<'ENDOFBODY'
Hey @sagarkhandagre998 — thanks for the work on this, migrating JSP pages to React is exactly the direction we need to go. I went through all 15 files and found some things that need fixing before we can merge. Grouping by priority:

## Blocking

1. **Hardcoded status ID `"22"` in reject endpoint** — use `StatusService.getStatusID(ExternalOrderStatus.Cancelled)` instead. Hardcoded IDs break across environments.

2. **Wrong service method** — controller calls `searchForElectronicOrders(form)` but the existing JSP controller uses `searchForStudyElectronicOrders(form)`. These have different logic.

3. **FACILITY search is broken end-to-end** — `FACILITY` was never added to the `SearchType` enum, neither service method handles it, and the frontend never renders the org dropdown. Either implement it fully or remove the dead code.

4. **~25 i18n keys missing from `en.json`** — all new `FormattedMessage` keys need to be added to `frontend/src/languages/en.json`. Project requirement.

5. **Liquibase filenames 023/024 collide** with files already on develop (`023-storage-device-connectivity.xml`, `024-freezer-monitoring-schema.xml`). Renumber to the next available slots.

6. **Remove `OpenApiFixController.java`** — unrelated to this feature and will break any future OpenAPI/Swagger setup.

7. **Cypress tests are deprecated** — new E2E tests need to use Playwright. See `AGENTS.md` for the test framework policy.

## High

8. **Patient identifiers in INFO logs** — `LogEvent.logInfo` with search values is a PII/compliance concern. Use DEBUG level or remove.

9. **Error response leaks internals** — `"Error: " + e.getMessage()` can expose SQL/stack traces. Return a generic message.

10. **`/rest/menu` breaking change** — swapping `getMenuTree()` for `getMenuTreeForReactUI()` on the existing endpoint removes items for other consumers. Use a query param or separate endpoint instead.

11. **Controller does FHIR calls directly** — `convertToDisplayItem` makes 3+ FHIR REST calls per order. This belongs in a service class, and it creates an N+1 performance issue.

## Minor

12. Use React Router navigation instead of `window.location.href` (breaks SPA)
13. Remove leftover `console.error` calls
14. Typo in migration: `tooltip.bannner` (triple-n)
15. Use Carbon spacing instead of `<br />` tags and inline styles

The core approach (menu filtering with `hideInNewUI`, component structure, route setup) is solid. Once the blocking items are addressed this will be in good shape.
ENDOFBODY
)

pmanko reviewed (changes requested):
Hey @sagarkhandagre998 — thanks for the work on this, migrating JSP pages to React is exactly the direction we need to go. I went through all 15 files and found some things that need fixing before we can merge. Grouping by priority:

## Blocking

1. **Hardcoded status ID `"22"` in reject endpoint** — use `StatusService.getStatusID(ExternalOrderStatus.Cancelled)` instead. Hardcoded IDs break across environments.

2. **Wrong service method** — controller calls `searchForElectronicOrders(form)` but the existing JSP controller uses `searchForStudyElectronicOrders(form)`. These have different logic.

3. **FACILITY search is broken end-to-end** — `FACILITY` was never added to the `SearchType` enum, neither service method handles it, and the frontend never renders the org dropdown. Either implement it fully or remove the dead code.

4. **~25 i18n keys missing from `en.json`** — all new `FormattedMessage` keys need to be added to `frontend/src/languages/en.json`. Project requirement.

5. **Liquibase filenames 023/024 collide** with files already on develop (`023-storage-device-connectivity.xml`, `024-freezer-monitoring-schema.xml`). Renumber to the next available slots.

6. **Remove `OpenApiFixController.java`** — unrelated to this feature and will break any future OpenAPI/Swagger setup.

7. **Cypress tests are deprecated** — new E2E tests need to use Playwright. See `AGENTS.md` for the test framework policy.

## High

8. **Patient identifiers in INFO logs** — `LogEvent.logInfo` with search values is a PII/compliance concern. Use DEBUG level or remove.

9. **Error response leaks internals** — `"Error: " + e.getMessage()` can expose SQL/stack traces. Return a generic message.

10. **`/rest/menu` breaking change** — swapping `getMenuTree()` for `getMenuTreeForReactUI()` on the existing endpoint removes items for other consumers. Use a query param or separate endpoint instead.

11. **Controller does FHIR calls directly** — `convertToDisplayItem` makes 3+ FHIR REST calls per order. This belongs in a service class, and it creates an N+1 performance issue.

## Minor

12. Use React Router navigation instead of `window.location.href` (breaks SPA)
13. Remove leftover `console.error` calls
14. Typo in migration: `tooltip.bannner` (triple-n)
15. Use Carbon spacing instead of `<br />` tags and inline styles

The core approach (menu filtering with `hideInNewUI`, component structure, route setup) is solid. Once the blocking items are addressed this will be in good shape.

sagarkhandagre998 commented:
Thanks a lot @pmanko for the review , i will address this all soon and kindly review this  https://github.com/DIGI-UW/OpenELIS-Global-2/pull/2862 migration also when you get a chance.

sagarkhandagre998 committed:
fixing blocking changes

1. Replace hardcoded status ID '22' with StatusService.getStatusID(ExternalOrderStatus.Cancelled) in reject endpoint
2. Fix wrong service method — controller now calls searchForStudyElectronicOrders instead of searchForElectronicOrders
3. Remove FACILITY dead code end-to-end — dropped organizationId/organizationOptions state, handleElectronicOrders, searchByFacility function, and the unused initial fetch from StudyEOrderSearch.js
4. Add 32 missing i18n keys to en.json covering eorder.reject.*, eorder.study.header, eorder.search.result.title, and all study.eorder.* keys
5. Renumber Liquibase files 023/024 to 028/029 to avoid collision with existing 023-storage-device-connectivity.xml and 024-freezer-monitoring-schema.xml on develop
6. Remove OpenApiFixController.java — unrelated to this feature and conflicts with future OpenAPI/Swagger setup

sagarkhandagre998 committed:
fixing high issues

1. Downgrade all LogEvent.logInfo calls that include search values or order IDs to logDebug to prevent PII leaking into INFO-level logs
2. Replace e.getMessage() in error ResponseEntity bodies with a generic 'An unexpected error occurred' message to prevent internal details leaking to clients
3. Restore GET /rest/menu to return the full MenuUtil.getMenuTree(); add ?view=react query param to serve the React-filtered tree, and update Header.js to use /rest/menu?view=react
4. Move convertToDisplayItem FHIR logic out of the controller into ElectronicOrderService — added buildStudyElectronicOrderDisplayItem to the interface and implemented it in ElectronicOrderServiceImpl; controller now delegates via service method reference

sagarkhandagre998 committed:
fixing minor issues

1. Replace window.location.href with useHistory().push() from react-router-dom to preserve SPA navigation
2. Remove console.error call from StudyEOrderSearch parseEOrders handler
3. Fix typo in Liquibase migration: tooltip.bannner -> tooltip.banner in 029-add-study-electronic-orders-react-menu.xml
4. Replace <br /> spacers and inline style={{ display: flex }} divs with Carbon Stack component (orientation and gap props) in StudyEOrder.js and StudyEOrderSearch.js

sagarkhandagre998 committed:
Merge develop into react-ui; resolve conflicts in Header.js, en.json, base.xml

- Header.js: kept develop refactored useEffect/loadHeaderLogo code, preserved ?view=react on /rest/menu call
- en.json: kept all react-ui new keys (eorder.reject.*, study.eorder.*, storage.*) and added develop barcode.labels.* keys
- base.xml: kept all develop includes; renumbered our Liquibase files 028->031 and 029->032 to avoid collision with develop's 028-create-site-branding-table.xml and 029-add-gps-coordinates-to-sample.xml

sagarkhandagre998 committed:
fix duplicate key

sagarkhandagre998 committed:
apply formatting

sagarkhandagre998 committed:
test: migrate StudyElectronicOrders from Cypress to Playwright

sagarkhandagre998 committed:
Merge upstream/develop; move spec to foundational/core; rename 031→033/034 liquibase files

sagarkhandagre998 committed:
fix(tests): update menu mock URL to match ?view=react query param

sagarkhandagre998 committed:
fix: validate page param before parseInt; return 400 on non-numeric or non-positive value

sagarkhandagre998 commented on frontend/cypress/e2e/studyElectronicOrders.cy.js:
```diff
@@ -0,0 +1,323 @@
+
+    // Close modal - just click cancel, don't verify close due to timing issues
+    cy.contains("button", "Cancel").click();
+    cy.wait(1000);
```
removed and migrated to playwright

sagarkhandagre998 committed:
removed unused

sagarkhandagre998 committed:
fix(security): map REST endpoints to StudyElectronicOrdersReact module for RBAC enforcement

sagarkhandagre998 committed:
fix(playwright): correct fixture import path for study-electronic-orders spec

sagarkhandagre998 commented:
@pmanko Now everything has fixed .

sagarkhandagre998 committed:
fix: remove duplicate keys from en.json

sagarkhandagre998 committed:
Merge branch 'develop' into react-ui

sagarkhandagre998 committed:
Rename 033-add-hide-in-new-ui-column.xml to 035-add-hide-in-new-ui-column.xml

sagarkhandagre998 committed:
Rename 034-add-study-electronic-orders-react-menu.xml to 036-add-study-electronic-orders-react-menu.xml

sagarkhandagre998 committed:
Merge remote-tracking branch 'upstream/develop' into react-ui

# Conflicts:
#	frontend/src/App.jsx
#	frontend/src/languages/en.json
#	src/main/resources/liquibase/3.3.x.x/base.xml

sagarkhandagre998 committed:
rename .js file to .jsx
